pypeit/core/moment.py

```diff
@@ -0,0 +1,497 @@
+        if _row.ndim == 2 and _col.ndim == 2 and _col.shape != _row.shape:
+            raise ValueError('col and row must have the same shape.')
+        if len(_row) == 1 and _col.ndim == 2 or _row.ndim == 2 and _col.ndim == 1:
+            raise ValueError('Invalid combination of row and col input shapes.')
+    else:
+        _row = np.arange(nrow)
+    # Check column and row values are valid
+
+    # TODO: This allows the window centers to be outside the image
+    # range. This is okay because _col is never used when slicing the
+    # image. However, _row values are, meaning that they have to be
+    # within the image limits.
+#    if np.any((_col < 0) | (_col >= ncol)):
+#        raise ValueError('Column locations outside provided image.')
+    if np.any((_row < 0) | (_row >= nrow)):
+        raise ValueError('Row locations outside provided image.')
+
+    # Fill the columns and rows so that they have matching shape
+    rmrowdim = _row.size == 1
+    rmcoldim = _col.size == 1 and not array_input
+    if _col.ndim == 1 and _row.ndim == 1:
+        if _row.size == 1 or _row.size == _col.size and not mesh:
+            if _row.size == 1:
+                # Repeat the row for each column
+                _row = np.full(_col.size, _row[0], dtype=int)
+            else:
+                # mesh is False, so just match column and row numbers
+                _row.reshape(1,-1)
+                rmrowdim = True
+            _col = _col.reshape(1,-1)
+            _width = _width.reshape(1,-1)
+        else:
+            # Do the combinatorics for rows, columns, and integration
+            # widths for the measurements
+            nmomcol = _col.size
+            _col = np.tile(_col, (_row.size,1))
+            _width = np.tile(_width, (_row.size,1))
+            _row = np.tile(_row, (nmomcol,1)).T
+    elif _col.ndim == 2 and _row.ndim == 1:
+        _row = np.tile(_row, (_col.shape[1],1)).T
+
+    # Construct the shape for the output
+    outshape = (norder,) + _col.shape
+    if rmrowdim and rmcoldim:
+        outshape = (outshape[0],)
+    elif rmrowdim:
+        outshape = (outshape[0],outshape[2])
+    elif rmcoldim:
+        outshape = (outshape[0],outshape[1])
+    if norder == 1 and len(outshape) > 1:
+        outshape = outshape[1:]
+
+    # Flatten the coordinate arrays (creates a copy).
+    _row = _row.flatten().astype(int)
+    _col = _col.flatten().astype(float)
+    _width = _width.flatten().astype(float)
+
+    # The "radius" of the pixels to cover is either half of the
+    # provided width for uniform weighting or 3*width for Gaussian
+    # weighting, where width is the sigma of the Gaussian
+    _radius = _width/2 if _weighting == 'uniform' else _width*3
+
+    # Window for the integration for each coordinate
+    i1 = np.floor(_col - _radius + 0.5).astype(int)
+    i2 = np.floor(_col + _radius + 0.5).astype(int)
+    c = i1[:,None]-1+np.arange(int(np.amax(np.amin(i2-i1)-1,0))+4)[None,:]
```

Comment:
Yes, this is a bit tricky and wasn't consistent between `trace_fweight`, `trace_gweight` and `extract_asymbox2`.  I tested this and I think it would be fine if it was just 2.  By setting to 4, there's a pixel with 0 weight at both ends of the integration window.